File: providers/medium-provider.ts
import Parser from "rss-parser";
import { Window } from "happy-dom";
import { IBlogPost, BlogPostProvider } from "../types/blog";
import { toKebabCase } from "../utils/string";

export class MediumProvider implements BlogPostProvider {
  private readonly MAX_RETRIES = 3;
  private readonly INITIAL_RETRY_DELAY = 1000;

  async getPosts(): Promise<IBlogPost[]> {
    let lastError: Error | null = null;
    
    for (let attempt = 0; attempt < this.MAX_RETRIES; attempt++) {
      try {
        const response = await fetch('https://medium.com/feed/@ilyasakin', {
          next: { revalidate: 86400 }, // Cache for 24 hours
        });
        
        if (!response.ok) {
          throw new Error(`HTTP error! status: ${response.status}`);
        }
        
        const xmlText = await response.text();
        const parser = new Parser();
        const feed = await parser.parseString(xmlText);

        return feed.items.map(this.transformPost);
      } catch (err) {
        lastError = err as Error;
        
        if (err instanceof Error && err.message.includes('429')) {
          const delay = this.INITIAL_RETRY_DELAY * Math.pow(2, attempt);
          console.log(`Received 429, retrying in ${delay}ms (attempt ${attempt + 1}/${this.MAX_RETRIES})`);
          await new Promise(resolve => setTimeout(resolve, delay));
          continue;
        }
        
        console.error("Failed to fetch Medium feed:", err);
        throw err;
      }
    }

    throw lastError || new Error("Unexpected error in getMediumPosts");
  }

  async getPost(slug: string): Promise<IBlogPost | null> {
    const posts = await this.getPosts();
    return posts.find(post => toKebabCase(post.title) === slug) || null;
  }

  private transformPost(item: any): IBlogPost {
    const window = new Window();
    const document = window.document;
    document.body.innerHTML = item["content:encoded"] || "";

    // Remove tracking images
    document.querySelectorAll("img").forEach((img) => {
      if (img.src.includes("post.clientViewed")) {
        img.remove();
      }
    });

    const originalInnerHTML = document.body.innerHTML;

    // Remove figures and headers for preview text
    document.querySelectorAll("figure").forEach((figure) => {
      figure.remove();
    });
    document.querySelectorAll("h1, h2, h3, h4, h5, h6").forEach((header) => {
      header.remove();
    });
    const textContent = document.body.textContent || "";

    return {
      title: item.title || "",
      slug: toKebabCase(item.title || ""),
      link: item.link || "",
      pubDate: item.pubDate || "",
      content: originalInnerHTML,
      preview: textContent.slice(0, 200).trim() + "...",
      thumbnail: item["content:encoded"]?.match(
        /<img[^>]*src="([^"]*)"[^>]*>/,
      )?.[1],
      source: 'medium'
    };
  }
} 

File: internal/services/medium_provider.go
package services

import (
	"context"
	"regexp"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/ilyasakin/ilyasakin.github.io/internal/models"
	"github.com/microcosm-cc/bluemonday"
	"github.com/mmcdole/gofeed"
)

type MediumProvider struct{}

func NewMediumProvider() *MediumProvider { return &MediumProvider{} }

func (p *MediumProvider) GetPosts(ctx context.Context) ([]*models.BlogPost, error) {
	parser := gofeed.NewParser()
	feed, err := parser.ParseURLWithContext("https://medium.com/feed/@ilyasakin", ctx)
	if err != nil {
		return nil, err
	}
	posts := make([]*models.BlogPost, 0, len(feed.Items))
	for _, it := range feed.Items {
		if it == nil {
			continue
		}
		content := it.Content
		if content == "" {
			content = it.Description
		}
		sanitized := sanitizeMediumHTML(content)
		preview := buildPreview(sanitized)
		posts = append(posts, &models.BlogPost{
			Title:   it.Title,
			Slug:    toKebabCase(it.Title),
			Link:    it.Link,
			PubDate: parsePubDate(it.PublishedParsed),
			Content: sanitized,
			Preview: preview,
			Source:  "medium",
		})
	}
	return posts, nil
}

func (p *MediumProvider) GetPost(ctx context.Context, slug string) (*models.BlogPost, error) {
	posts, err := p.GetPosts(ctx)
	if err != nil {
		return nil, err
	}
	for _, post := range posts {
		if post.Slug == slug {
			return post, nil
		}
	}
	return nil, nil
}

func parsePubDate(t *time.Time) time.Time {
	if t == nil {
		return time.Now()
	}
	return *t
}

func sanitizeMediumHTML(html string) string {
	// Remove tracking images
	re := regexp.MustCompile(`<img[^>]*post.clientViewed[^>]*>`)
	cleaned := re.ReplaceAllString(html, "")
	// Basic policy allowing common content tags
	policy := bluemonday.UGCPolicy()
	policy.AllowElements("figure", "figcaption")
	policy.AllowAttrs("src", "alt", "width", "height", "class").OnElements("img")
	return policy.Sanitize(cleaned)
}

func buildPreview(html string) string {
	// Work on a DOM to selectively remove figures and headers
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		return ""
	}
	// Remove tracking images if any survived
	doc.Find("img").Each(func(_ int, s *goquery.Selection) {
		if src, exists := s.Attr("src"); exists && strings.Contains(src, "post.clientViewed") {
			s.Remove()
		}
	})
	// Remove figures and headers for preview text
	doc.Find("figure, h1, h2, h3, h4, h5, h6").Remove()
	text := strings.TrimSpace(doc.Text())
	if len(text) > 200 {
		text = text[:200] + "..."
	}
	return text
}

func toKebabCase(s string) string {
	s = strings.ToLower(s)
	s = strings.NewReplacer("(", " ", ")", " ").Replace(s)
	re := regexp.MustCompile(`[^a-z0-9]+`)
	s = re.ReplaceAllString(s, "-")
	s = strings.Trim(s, "-")
	return s
}
